pkg/pillar/cmd/zedagent/handlecertconfig.go

```diff
@@ -0,0 +1,152 @@
+}
+
+// controller cert update, triggers cipher context update
+func handleControllerCertUpdate(ctx *getconfigContext,
+	controllerCert types.ControllerCertificate, reset bool) {
+	//switch controllerCert.Type {
+	//case zcert.ZCertType_CERT_TYPE_CONTROLLER_ECDH_EXCHANGE:
+	//	updateCipherContextsWithControllerCert(ctx, controllerCert, reset)
+	//	return
+	//case zcert.ZCertType_CERT_TYPE_CONTROLLER_SIGNING:
+	//case zcert.ZCertType_CERT_TYPE_CONTROLLER_INTERMEDIATE:
+	//	// TBD:XXX add appropriate handlers
+	//}
+	updateCipherContextsWithControllerCert(ctx, controllerCert, reset)
+	return
+
+}
+
+// update the cipher context with the controller cert
+func updateCipherContextsWithControllerCert(ctx *getconfigContext,
+	controllerCert types.ControllerCertificate, reset bool) {
+	log.Infof("%v, update cipher contexts\n", controllerCert.Key())
+	items := ctx.pubCipherContextConfig.GetAll()
+	for _, item := range items {
+		cipherCtx := item.(types.CipherContext)
+		// hash algorithm do not match
+		//if cipherCtx.HashScheme.String() !=
+		//	controllerCert.HashAlgo.String() ||
+		//	!bytes.Equal(cipherCtx.ControllerCertHash,
+		//		controllerCert.CertHash) {
+		//	continue
+		//}
+		if !bytes.Equal(cipherCtx.ControllerCertHash,
+			controllerCert.CertHash) {
+			continue
+		}
+		log.Infof("%v, updating ciphercontext, %s\n",
+			controllerCert.Key(), cipherCtx.Key())
+		if reset {
+			if len(cipherCtx.ControllerCert) != 0 {
+				cipherCtx.ControllerCert = []byte{}
+				publishCipherContextConfig(ctx, cipherCtx)
+			}
+			continue
+		}
+		cipherCtx.ControllerCert = controllerCert.Cert
+		publishCipherContextConfig(ctx, cipherCtx)
+	}
+}
+
+func getControllerCert(ctx *zedagentContext, suppliedHash []byte) []byte {
+	log.Infof("%v, get controller cert\n", suppliedHash)
+	items := ctx.subControllerCertConfig.GetAll()
+	for _, item := range items {
+		certConfig := item.(types.ControllerCertificate)
+		if bytes.Equal(certConfig.CertHash, suppliedHash) {
+			return certConfig.Cert
+		}
+	}
+	return []byte{}
+}
+
+func publishControllerCertConfig(ctx *getconfigContext,
+	config types.ControllerCertificate) {
+	key := config.Key()
+	log.Debugf("publishControllerCertificate %s\n", key)
+	pub := ctx.pubControllerCertConfig
+	pub.Publish(key, config)
+}
+
+func unpublishControllerCertConfig(ctx *getconfigContext, key string) {
+	log.Debugf("unpublishControllerCertConfig %s\n", key)
+	pub := ctx.pubControllerCertConfig
+	c, _ := pub.Get(key)
+	if c == nil {
+		log.Errorf("unpublishCertObjConfig(%s) not found\n", key)
+		return
+	}
+	pub.Unpublish(key)
+}
+
+// for device cert
+func getDeviceCert(hashScheme zconfig.CipherHashAlgorithm,
+	suppliedHash []byte) []byte {
+	log.Infof("%v, get device cert\n", suppliedHash)
+	// TBD:XXX as of now, only one
+	certBytes, err := ioutil.ReadFile(types.DeviceCertName)
+	if err == nil {
+		if computeAndMatchHash(certBytes, suppliedHash, hashScheme) {
+			return certBytes
+		}
+	}
+	return []byte{}
```

Comment:
What do other agents do when they can't read the device cert? I suspect they all log.Error. They might even do more drastical things. But silence like here sounds quite bad.

Reply:
This is now part of orchestration. The error will be posted to controller.